builtin/common/vector.lua

```diff
@@ -141,3 +141,95 @@ function vector.sort(a, b)
 	return {x = math.min(a.x, b.x), y = math.min(a.y, b.y), z = math.min(a.z, b.z)},
 		{x = math.max(a.x, b.x), y = math.max(a.y, b.y), z = math.max(a.z, b.z)}
 end
+
+local function sin(x)
+	if x % math.pi == 0 then
+		return 0
+	else
+		return math.sin(x)
+	end
+end
+
+local function cos(x)
+	if x % math.pi == math.pi / 2 then
+		return 0
+	else
+		return math.cos(x)
+	end
+end
+
+function vector.rotate_around_axis(v, axis, angle)
+	local cosangle = cos(angle)
+	local sinangle = sin(angle)
+	axis = vector.normalize(axis)
+	--https://en.wikipedia.org/wiki/Rodrigues%27_rotation_formula
+	local dot_axis = vector.multiply(axis, vector.dot(axis, v))
+	return vector.add(
+		vector.multiply(vector.cross(axis, v), sinangle),
+		vector.add(
+		vector.multiply(vector.subtract(v, dot_axis), cosangle),
+		dot_axis
+	))
+end
+
+function vector.rotate(v, rot)
+	local sinpitch = sin(-rot.x)
+	local sinyaw = sin(-rot.y)
+	local sinroll = sin(-rot.z)
+	local cospitch = cos(rot.x)
+	local cosyaw = cos(rot.y)
+	local cosroll = math.cos(rot.z)
+	--rotation matrix that applies yaw, pitch and roll
+	--result of matrix multiplying the rotation matrices for yaw, pitch and roll
+	local matrix = {
+		{
+			cosyaw * cospitch,
+			cosyaw * sinpitch * sinroll - sinyaw * cosroll,
+			cosyaw * sinpitch * cosroll + sinyaw * sinroll
+		},
+		{
+			sinyaw * cospitch,
+			sinyaw * sinpitch * sinroll + cosyaw * cosroll,
+			sinyaw * sinpitch * cosroll - cosyaw * sinroll
+		},
+		{
+			-sinpitch,
+			cospitch * sinroll,
+			cospitch * cosroll
+		},
+	}
+	local keys = {"z", "x", "y"}
+	--compute matrix multiplication: `matrix` * `v`
+	local ret = vector.new(0, 0, 0)
+	for i = 1, 3 do
+		local row = matrix[i]
+		for ii = 1, 3 do
+			ret[keys[i]] = ret[keys[i]] + row[ii] * v[keys[ii]]
+		end
+	end
+	return ret
+end
+
+
+function vector.forward_at_rotation(rot)
+	return vector.rotate({x = 0, y = 0, z = 1}, rot)
+end
+
+function vector.up_at_rotation(rot)
+	return vector.rotate({x = 0, y = 1, z = 0}, rot)
+end
```

Comment:
These are so trivial I don't think we need them in the API.
For `vector.forward_at_rotation` the implication that Z is "forward" can also be problematic.